---
title: "Privratsky-20210215 - Clustering resolution 0.5 / 23 clusters"
author: "Hélène Fradin Kirshner"
date: "03/15/2021"
output: html_document
---

source conda/bin/activate privratsky
rmarkdown::render('./5_Clustering_res_0.5/5_Clustering_res_0.5.Rmd')

Changes in myeloid and kidney cells after CLP - Analysis of 2 x 10X scRNA-seq samples from 2 pools of WT mice (3 Sham + 3 CLP): comparison of gene expression in different cell populations
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=TRUE, warning=FALSE, message=FALSE, error=FALSE
                      , tidy.opts=list(width.cutoff=60),tidy=TRUE
                      , cache=TRUE)
knitr::opts_knit$set(root.dir = '/data/omicscore/Privratsky-Privratsky-20210215')
# knitr::opts_knit$set(root.dir = '~/Volumes/HARDAC-OmicsCore/Privratsky-Privratsky-20210215')
```

```{r}
rainbow1.7c <- c("#B3ADD3", "#80617D", "#A75769", "#F9C971", "#A3C587", "#40A8AC", "#296D71")
slices <- rep(1, length(rainbow1.7c))
pie(slices, col=rainbow1.7c)
```

```{r}
rainbow2.6c <- c("#03539C", "#12999E", "#B7CE05", "#FAEB09", "#FDA908", "#E82564")
slices <- rep(1, length(rainbow2.6c))
pie(slices, col=rainbow2.6c)
```

```{r}
rainbow3.5c <- c("#f66e6e", "#f6b36e", "#f5f66e", "#6ef3f6", "#9c6ef6")
slices <- rep(1, length(rainbow3.5c))
pie(slices, col=rainbow3.5c)
```

```{r}
rainbow4.12c <- c("#1a1334", "#27294a", "#01545a", "#017352", "#02c383", "#abd962", "#fbbf46", "#ef6b32", "#ee0445", "#a22b5e", "#710062", "#022c7d")
slices <- rep(1, length(rainbow4.12c))
pie(slices, col=rainbow4.12c)
```

```{r}
indir <- "./processedData/2_1_Resolution_choice"
outdir <- "./processedData/5_Clustering_res_0.5"
dir.create(outdir, recursive = T)
```

```{r}
library(Seurat)
integrated <- readRDS(paste0(indir, "/8.integrated.rds"))
integrated
```

```{r}
Idents(integrated) <- "integrated_snn_res.0.5"
table(integrated@active.ident)
```

```{r}
# pal <- colorRampPalette(c("#12999E", "#FAEB09", "#E82564", "#03539C"))
pal <- colorRampPalette(rainbow1.7c)
levels <- levels(integrated$integrated_snn_res.0.5)
colors.clusters <- pal(length(levels))
colors.clusters.2 <- colors.clusters[c(11:14, 1, 7, 2, 20, 3:4, 8, 5, 9, 15, 6, 10, 16, 21, 17:18, 22:23, 19)]
names(colors.clusters.2) <- levels
colors.clusters.2
slices <- rep(1,length(levels))
pie(slices, col=colors.clusters.2, labels = names(colors.clusters.2))
```

```{r}
d <- DimPlot(integrated, reduction = "umap", pt.size = 0.2, label = T, label.size = 6,
             cols = colors.clusters.2)
d
pdf(paste0(outdir, "/1_DimPlot_umap_clusters_pc50_res0.5.pdf")
    , width = 10, height = 8)
d
dev.off()
```

```{r}
colors.samples <- c("#12999E", "#FDA908")
names(colors.samples) <- levels(as.factor(integrated$sample.id))
slices <- rep(1,length(colors.samples))
pie(slices, col=colors.samples, labels = names(colors.samples))
```

```{r}
p1 <- DimPlot(integrated, reduction = "umap", group.by = "sample.id", pt.size = 0.2, cols = colors.samples)
p2 <- DimPlot(integrated, reduction = "umap", label = TRUE, pt.size = 0.2, label.size = 6, cols = colors.clusters.2)
library(cowplot)
plot_grid(p1, p2)
pdf(paste0(outdir, "/2_2DimPlots_umap_samples_clusters_pc50_res0.5.pdf")
    , width = 18, height = 8)
plot_grid(p1, p2)
dev.off()
```

```{r}
d <- DimPlot(integrated, reduction = "umap", group.by = "sample.id", split.by = "sample.id", pt.size = 0.2, ncol = 2,
             cols = colors.samples)
d
pdf(paste0(outdir, "/3_DimPlot_umap_split_by_samples.pdf"),
    width = 16, height = 9)
d
dev.off()
```

```{r}
f <- FeaturePlot(integrated, features = c("Nphs2", "Slc5a2", "Clcnka", "Slc12a1", "Ptgs2", "Slc12a3", "Calb1", "Aqp2", "Slc4a1", "Slc26a4", "Slc14a2", "Upk1a", "Cd22", "Adgre1" , "Pecam1", "Pdgfrb", "Cd68", "Cd14", "Acta2", "Csf3r", "Cd4"), min.cutoff = "q9"
                 , order = T)
f
pdf(paste0(outdir, "/4_FeaturePlot_cellID.pdf")
    , width = 28, height = 42)
f
dev.off()
```
##Annotation of markers based on cluster markers from Susztak Science paper (Park et al., Science 360, 758–763 (2018) and Kidney International (2019) 95, 787–796; https://doi.org/10.1016/

https://science.sciencemag.org/content/360/6390/758.long
https://www.kidney-international.org/article/S0085-2538(18)30912-8/fulltext

#Podocyte markers -> cluster 28
```{r}
f2 <- FeaturePlot(integrated, features = c("Nphs2", "Podxl"), min.cutoff = "q9")
f2
pdf(paste0(outdir, "/5_FeaturePlot_Podo.pdf")
    , width = 14, height = 7)
f2
dev.off()
```

#Endothelial markers -> cluster 15
```{r}
f3 <- FeaturePlot(integrated, features = c("Plat", "Pecam1"), min.cutoff = "q9")
f3
pdf(paste0(outdir, "/6_FeaturePlot_Endo.pdf")
    , width = 14, height = 7)
f3
dev.off()
```

#PT-S1 markers -> clusters 7,8,9
```{r}
f4 <- FeaturePlot(integrated, features = c("Slc5a2", "Slc5a12"), min.cutoff = "q9")
f4
pdf(paste0(outdir, "/7_FeaturePlot_PTs1.pdf")
    , width = 14, height = 7)
f4
dev.off()
```

#PT-S2 markers
```{r}
f5 <- FeaturePlot(integrated, features = c("Fxyd2", "Hrsp12"), min.cutoff = "q9")
f5
pdf(paste0(outdir, "/8_FeaturePlot_PTs2.pdf")
    , width = 7, height = 7)
f5
dev.off()
```

#PT-S3 markers -> cluster 5
```{r}
f6 <- FeaturePlot(integrated, features = c("Atp11a", "Slc13a3"), min.cutoff = "q9")
f6
pdf(paste0(outdir, "/9_FeaturePlot_PTs3.pdf")
    , width = 14, height = 7)
f6
dev.off()
```

#Loop of Henle -> clusters 11, 13, 18
```{r}
f7 <- FeaturePlot(integrated, features = c("Slc12a1", "Umod"), min.cutoff = "q9")
f7
pdf(paste0(outdir, "/10_FeaturePlot_LOH.pdf")
    , width = 14, height = 7)
f7
dev.off()
```

#Distal CT -> cluster 10
```{r}
f8 <- FeaturePlot(integrated, features = c("Slc12a3", "Pvalb"), min.cutoff = "q9")
f8
pdf(paste0(outdir, "/11_FeaturePlot_DCT.pdf")
    , width = 14, height = 7)
f8
dev.off()
```

#Conn Tubule -> clusters 6, 20, 21, 29
```{r}
f21 <- FeaturePlot(integrated, features = c("Calb1"), min.cutoff = "q9")
f21
pdf(paste0(outdir, "/12_FeaturePlot_ConnTub.pdf")
    , width = 7, height = 7)
f21
dev.off()
```

#CD PC -> cluster 21
```{r}
f9 <- FeaturePlot(integrated, features = c("Aqp2", "Hsd11b2"), min.cutoff = "q9")
f9
pdf(paste0(outdir, "/13_FeaturePlot_CD-PC.pdf")
    , width = 14, height = 7)
f9
dev.off()
```

#CD-IC -> clusters 24, 29, 39
```{r}
f10 <- FeaturePlot(integrated, features = c("Atp6v1g3", "Atp6v0d2"), min.cutoff = "q9")
f10
pdf(paste0(outdir, "/14_FeaturePlot_CD-IC.pdf")
    , width = 14, height = 7)
f10
dev.off()
```

#CD Trans -> cluster 29
```{r}
f11 <- FeaturePlot(integrated, features = c("Slc26a4", "Insrr", "Rhbg"), min.cutoff = "q9")
f11
pdf(paste0(outdir, "/15_FeaturePlot_CD-Trans.pdf")
    , width = 14, height = 14)
f11
dev.off()
```

#Fibroblast
```{r}
f12 <- FeaturePlot(integrated, features = c("Plac8", "S100a4", "Pdgfrb"), min.cutoff = "q9")
f12
pdf(paste0(outdir, "/16_FeaturePlot_Fib.pdf")
    , width = 14, height = 14)
f12
dev.off()
```

#Macro -> cluster 22
```{r}
f13 <- FeaturePlot(integrated, features = c("C1qa", "Cd68", "C1qb"), min.cutoff = "q9")
f13
pdf(paste0(outdir, "/17_FeaturePlot_Macro.pdf")
    , width = 14, height = 14)
f13
dev.off()
```

#PMN -> cluster 36
```{r}
f14 <- FeaturePlot(integrated, features = c("S100a8", "Ly6g", "S100a9"), min.cutoff = "q9")
f14
pdf(paste0(outdir, "/18_FeaturePlot_PMN.pdf")
    , width = 14, height = 14)
f14
dev.off()
```

#B lymph -> cluster 37
```{r}
f15 <- FeaturePlot(integrated, features = c("Cd79a", "Cd79b", "Cd19"), min.cutoff = "q9")
f15
pdf(paste0(outdir, "/19_FeaturePlot_Blymph.pdf")
    , width = 14, height = 14)
f15
dev.off()
```

#Tlymph -> cluster 30
```{r}
f16 <- FeaturePlot(integrated, features = c("Ltb", "Cd4", "Cxcr6"), min.cutoff = "q9")
f16
pdf(paste0(outdir, "/20_FeaturePlot_Tlymph.pdf")
    , width = 14, height = 14)
f16
dev.off()
```

#NK -> cluster 30
```{r}
f17 <- FeaturePlot(integrated, features = c("Gzma", "Nkg7"), min.cutoff = "q9")
f17
pdf(paste0(outdir, "/21_FeaturePlot_NK.pdf")
    , width = 14, height = 7)
f17
dev.off()
```

#Novel1
```{r}
f18 <- FeaturePlot(integrated, features = c("Slc27a2", "Lrp2", "Cdca3"), min.cutoff = "q9")
f18
pdf(paste0(outdir, "/22_FeaturePlot_Novel1.pdf")
    , width = 14, height = 14)
f18
dev.off()
```

```{r}
#library(Seurat)
DefaultAssay(integrated) <- "RNA"
clusters <- levels(integrated@active.ident)
conserved.markers <- data.frame(matrix(ncol = 14))
 for (c in clusters) {
  print(c)
  markers.c <- FindConservedMarkers(integrated, ident.1 = c, grouping.var = "sample.id", verbose = T
                                    , logfc.threshold = -Inf
                                    , min.pct = -Inf
                                    , min.cells.feature = -Inf
                                    , min.cells.group = -Inf)
  markers.c <- cbind(data.frame(cluster = rep(c, dim(markers.c)[1])
                                , gene = rownames(markers.c)), markers.c)
  write.table(markers.c, file = paste0(outdir, "/23_markers_", c, ".txt"))
  colnames(conserved.markers) <- colnames(markers.c)
  conserved.markers <- rbind(conserved.markers, markers.c)
  head(conserved.markers)
}
conserved.markers <- conserved.markers[-1, ]
```

```{r}
conserved.markers <- conserved.markers[, c(1:2, 13:14, 3:12)]
head(conserved.markers)
```

Only top markers that are positive markers
```{r}
conserved.markers$marker.type <- apply(conserved.markers, 1, function(x) {
  y <- as.numeric(x)
  if ( (if (is.na(y[6])) {TRUE} else {y[6]>0})
       & (if (is.na(y[11])) {TRUE} else {y[11]>0})
       # & (if (is.na(y[16])) {TRUE} else {y[16]>0})
       # & (if (is.na(y[21])) {TRUE} else {y[21]>0})
       # & (if (is.na(y[26])) {TRUE} else {y[26]>0})
       # & (if (is.na(y[31])) {TRUE} else {y[31]>0})
       # & (if (is.na(y[36])) {TRUE} else {y[36]>0})
       # & (if (is.na(y[41])) {TRUE} else {y[41]>0}) 
       )
    {"POS"}
  else if ( ( if (is.na(y[6])) {TRUE} else {y[6]<0})
       & (if (is.na(y[11])) {TRUE} else {y[11]<0})
       # & (if (is.na(y[16])) {TRUE} else {y[16]<0})
       # & (if (is.na(y[21])) {TRUE} else {y[21]<0})
       # & (if (is.na(y[26])) {TRUE} else {y[26]<0})
       # & (if (is.na(y[31])) {TRUE} else {y[31]<0})
       # & (if (is.na(y[36])) {TRUE} else {y[36]<0})
       # & (if (is.na(y[41])) {TRUE} else {y[41]<0})
       ) 
    {"NEG"}
  else {"UND"}
  })
conserved.markers <- conserved.markers[, c(1:4, 15, 5:14)]
```

```{r}
openxlsx::write.xlsx(conserved.markers, paste0(outdir, "/23_conserved_markers_PC50_res0.5_23clusters.xlsx")
                     , colNames = T)
head(conserved.markers)
```

```{r}
library(Seurat)
x <- factor(integrated$integrated_snn_res.0.5)
head(x)
```

```{r}
levels(x)
```

```{r}
library(plyr)
x <- revalue(x, c(`0` = "PCT-S1"
                  , `1` = "PST-S3"
                  , `2` = "PT-S2"
                  , `3` = "PST-S3"
                  , `4` = "LOH"
                  , `5` = "Endo"
                  , `6` = "Conn Tubule"
                  , `7` = "Macro"
                  , `8` = "Distal CT"
                  , `9` = "CD-IC"
                  , `10` = "Endo"
                  , `11` = "CD-PC"
                  , `12` = "Fib"
                  , `13` = "PCT-S1"
                  , `14` = "CD Trans"
                  , `16` = "Fib"
                  , `17` = "NK"
                  , `18` = "Podo"
                  , `19` = "Fib"
                  , `20` = "Fib"
                  , `21` = "B lymph"
                  , `22` = "PMN"
                  ))
integrated$annotation.1 <- x
head(integrated$annotation.1)
```

```{r}
Idents(integrated) <- "annotation.1"
```

```{r}
library(ggsci)
levels <- levels(integrated$annotation.1)
colors.annotation.1 <- pal_ucscgb("default", alpha = 1)(26)[1:length(levels)]
names(colors.annotation.1) <- levels
colors.annotation.1
slices <- rep(1,length(levels))
pie(slices, col=colors.annotation.1, labels = names(colors.annotation.1))
```

```{r}
library(Seurat)
d2 <- DimPlot(integrated, label = TRUE, label.size = 4
              , group.by = "annotation.1"
              , cols = colors.annotation.1
              , repel = T)
d2
pdf(paste0(outdir, "/24_Dimplot_newidents.pdf")
    , width = 13
    , height = 9)
d2
dev.off()

```

```{r}
d3 <- DimPlot(integrated, group.by = "sample.id", split.by = "sample.id", pt.size = 0.2, ncol = 2
              , cols = colors.samples)
d3
pdf(paste0(outdir, "/25_DimPlot_newidents_split_by_samples.pdf"),
    width = 16, height = 9)
d3
dev.off()
```

# Identify cells expressing Il6
```{r}
DefaultAssay(integrated) <- "RNA"
f19 <- FeaturePlot(integrated, features = "Il6", order = T, label = T, label.size = 3)
f19
pdf(paste0(outdir, "/26_FeaturePlot_Il6.pdf")
    , width = 11
    , height = 10)
f19
dev.off()
```

```{r}
f20 <- FeaturePlot(integrated, features = c("Il6"), split.by = "sample.id", max.cutoff = 3, 
    cols = c("grey", "red")
    , order = T)
f20
pdf(paste0(outdir, "/27_FeaturePlot_Il6-sham-CLP.pdf")
    , width = 19
    , height = 10)
f20
dev.off()
```

```{r}
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
integrated$celltype.stim <- paste(Idents(integrated), integrated$sample.id, sep = "_")
integrated$celltype <- Idents(integrated)
Idents(integrated) <- "celltype"

```

```{r}
plots <- VlnPlot(integrated, features = c("Il6"), split.by = "sample.id", group.by = "celltype", 
    pt.size = 0, combine = FALSE)
library(patchwork)
wrap_plots(plots = plots, ncol = 1)
```

```{r}
d <- DotPlot(integrated, features = "Il6", group.by = "celltype.stim")
openxlsx::write.xlsx(d$data, paste0(outdir, "/28_IL6_expn_per_celltype_stim.xlsx"))
```

```{r}
d
pdf(paste0(outdir, "/29_DotPlot_IL6_celltype_stim.pdf")
    , width = 5
    , height = 9)
d
dev.off()
```

```{r}
Idents(integrated) <- "integrated_snn_res.0.5"
cluster19 <- WhichCells(integrated, idents = "19")
# others <- WhichCells(integrated, idents = "33", invert = T)
d <- DimPlot(integrated, label=T, group.by="celltype", cells.highlight= list(cluster19), cols.highlight = c("darkblue"
                                                                                                       # , "darkred"
                                                                                                       ), cols= "grey")
d
pdf(paste0(outdir, "/30_DimPlot_integrated_label_group.by_celltype_cell.highlight_cluster19.pdf"))
d
dev.off()
```

```{r}
saveRDS(integrated, paste0(outdir, "/31.integrated.rds"))
```

```{r}
integrated$res.0.5.stim <- paste(integrated$integrated_snn_res.0.5 , integrated$sample.id, sep = "_")

```

```{r}
d <- DotPlot(integrated, features = "Il6", group.by = "res.0.5.stim")
openxlsx::write.xlsx(d$data, paste0(outdir, "/32_IL6_expn_per_cluster_stim.xlsx"))
```

```{r}
Idents(integrated) <- "integrated_snn_res.0.5"
f20 <- FeaturePlot(integrated, features = c("Il6"), split.by = "sample.id", max.cutoff = 3, 
    cols = c("grey", "red")
    , order = T
    , label = T)
f20
pdf(paste0(outdir, "/33_FeaturePlot_Il6-sham-CLP_w_labels.pdf")
    , width = 19
    , height = 10)
f20
dev.off()
```

# Updated cell types

```{r}
library(Seurat)
x <- factor(integrated$integrated_snn_res.0.5)
head(x)
```

```{r}
levels(x)
```

```{r}
library(plyr)
x <- revalue(x, c(`0` = "PT-S1"
                  , `1` = "PT-S3"
                  , `2` = "PT-S2"
                  , `3` = "PT-S3"
                  , `4` = "LOH"
                  , `5` = "Endo"
                  , `6` = "Conn Tubule"
                  , `7` = "Macro"
                  , `8` = "Distal Conv T"
                  , `9` = "CD-IC"
                  , `10` = "Endo"
                  , `11` = "CD-PC"
                  , `12` = "Fib"
                  , `13` = "PT-S1"
                  , `14` = "CD-Trans"
                  , `15` = "CD-IM"
                  , `16` = "Fib"
                  , `17` = "NK"
                  , `18` = "Podo"
                  , `19` = "Fib"
                  , `20` = "Fib"
                  , `21` = "B lymph"
                  , `22` = "PMN"
                  ))
integrated$annotation.1 <- x
head(integrated$annotation.1)
```

```{r}
Idents(integrated) <- "annotation.1"
```

```{r}
cell.types <- c("Endo", "Podo", "PT-S1", "PT-S2", "PT-S3"
                , "LOH", "Distal Conv T", "Conn Tubule", "CD-PC", "CD-IC", "CD-Trans", "CD-IM"
                , "Fib", "Macro", "PMN", "B lymph", "NK")
```

```{r}
integrated$annotation.1 <- factor(integrated$annotation.1, levels = cell.types)
```

```{r}
library(ggsci)
levels <- levels(integrated$annotation.1)
colors.annotation.1 <- pal_ucscgb("default", alpha = 1)(26)[1:length(levels)]
names(colors.annotation.1) <- levels
colors.annotation.1
slices <- rep(1,length(levels))
pie(slices, col=colors.annotation.1, labels = names(colors.annotation.1))
```

```{r}
library(Seurat)
d2 <- DimPlot(integrated, label = TRUE, label.size = 4
              , group.by = "annotation.1"
              , cols = colors.annotation.1
              , repel = T)
d2
pdf(paste0(outdir, "/34_Dimplot_updated_newidents.pdf")
    , width = 13
    , height = 9)
d2
dev.off()

```

```{r}
library(Seurat)
Idents(integrated) <- "annotation.1"
d2 <- DimPlot(integrated, label = FALSE
              , cols = colors.annotation.1) + NoLegend()
d2
pdf(paste0(outdir, "/35_Dimplot_updated_newidents_NoLegend.pdf")
    , width = 7
    , height = 7)
d2
dev.off()

```

# DotPlot with cell markers

Endothelial	Plat
Endothelial	Pecam1
Podocytes	Nphs2
Podocytes	Podxl
PT-S1	Slc5a2
PT-S1	Slc5a12
PT-S2	Fxyd2
PT-S3	Atp11a
PT-S3	Slc13a3
LOH	Slc12a1
LOH	Umod
Distal CT	Slc12a3
Distal CT	Pvalb
Conn Tubule	Calb1
CD PC	Aqp2
CD PC	Hsd11b2
CD-IC	Atp6v1g3
CD-IC	Atp6v0d2
CD Trans	Slc26a4
CD Trans	Insrr
CD Trans	Rhbg
Inner medullary collecting duct (CD-IM)	Slc14a2
Fibroblast	Plac8
Fibroblast	S100a4
Fibroblast	Pdgfrb
Macro	C1qa
Macro	Cd68
Macro	C1qb
PMN	S100a8
PMN	Ly6g
PMN	S100a9
B lymph	Cd79a
B lymph	Cd79b
B lymph	Cd19
NK	Gzma
NK	Nkg7

```{r}
known.markers <- c("Plat", "Pecam1"
                   , "Nphs2", "Podxl"
                   , "Slc5a2", "Slc5a12"
                   , "Fxyd2"
                   , "Atp11a", "Slc13a3"
                   , "Slc12a1", "Umod"
                   , "Slc12a3", "Pvalb"
                   , "Calb1"
                   , "Aqp2", "Hsd11b2"
                   , "Atp6v1g3", "Atp6v0d2"
                   , "Slc26a4", "Insrr", "Rhbg"
                   , "Slc14a2"
                   , "Plac8", "S100a4", "Pdgfrb"
                   , "C1qa", "Cd68", "C1qb"
                   , "S100a8", "Ly6g", "S100a9"
                   , "Cd79a", "Cd79b", "Cd19"
                   , "Gzma", "Nkg7")
```

```{r}
DefaultAssay(integrated) <- "RNA"
d <- DotPlot(object = integrated
        , features =known.markers
        , cols = c("#03539C", "#E82564")
        , dot.scale = 8
        , group.by = "annotation.1"
          ) + RotatedAxis()
d
pdf(paste0(outdir, "/36_DotPlot_known_markers_cell_types.pdf")
    , width = 14
    , height = 7)
d
dev.off()
```

```{r}
saveRDS(integrated, paste0(outdir, "/37.annotated.rds"))
```

## Session Information
```{r}
sessionInfo()
```

```{r}
writeLines(capture.output(sessionInfo()), "./scripts/5_Clustering_res_0.5/5_Clustering_res_0.5.sessionInfo.txt")
```
